go/pkg/cas/upload.go

```diff
@@ -441,12 +476,67 @@ func (u *uploader) check(ctx context.Context, items []*uploadItem) error {
 }
 
 func (u *uploader) scheduleUpload(ctx context.Context, item *uploadItem) error {
-	if u.testScheduleUpload != nil {
-		return u.testScheduleUpload(ctx, item)
+	// Check if this blob can be uploaded in a batch.
+	if marshalledRequestSize(item.Digest) > int64(u.batchBundler.BundleByteLimit) {
+		// There is no way this blob can fit in a batch request.
+		// TODO(nodir): implement streaming.
+		panic("not implemented")
 	}
 
-	// TODO(nodir): implement.
-	panic("not implemented")
+	// Since this blob is small enough, just read it entirely.
+	contents, err := item.ReadAll()
+	if err != nil {
+		return errors.Wrapf(err, "failed to read the item")
+	}
+	req := &repb.BatchUpdateBlobsRequest_Request{Digest: item.Digest, Data: contents}
+	return u.batchBundler.AddWait(ctx, req, proto.Size(req))
+}
+
+// uploadBatch uploads blobs in using BatchUpdateBlobs RPC.
+func (u *uploader) uploadBatch(ctx context.Context, reqs []*repb.BatchUpdateBlobsRequest_Request) error {
+	if err := u.semBatches.Acquire(ctx, 1); err != nil {
+		return err
+	}
+	defer u.semBatches.Release(1)
+
+	reqMap := make(map[digest.Digest]*repb.BatchUpdateBlobsRequest_Request, len(reqs))
+	for _, r := range reqs {
+		reqMap[digest.NewFromProtoUnvalidated(r.Digest)] = r
+	}
+
+	req := &repb.BatchUpdateBlobsRequest{
+		InstanceName: u.InstanceName,
+		Requests:     reqs,
+	}
+	return u.retry(ctx, func() error {
+		// TODO(nodir): add per-RPC timeouts.
+		res, err := u.cas.BatchUpdateBlobs(ctx, req)
+		if err != nil {
+			return err
+		}
+
+		bytesTransferred := int64(0)
+		digestsTransferred := int64(0)
+		var retriableErr error
+		req.Requests = req.Requests[:0] // reset for the next attempt
+		for _, r := range res.Responses {
+			if err := status.FromProto(r.Status).Err(); err != nil {
+				if !retry.TransientOnly(err) {
+					return err
+				}
+				// This error is retriable. Save it to return later, and
+				// save the failed sub-request for the next attempt.
+				retriableErr = err
+				req.Requests = append(req.Requests, reqMap[digest.NewFromProtoUnvalidated(r.Digest)])
+				continue
+			}
+			bytesTransferred += r.Digest.SizeBytes
+			digestsTransferred++
+		}
+		atomic.AddInt64(&u.stats.Batched.Bytes, bytesTransferred)
```

Comment:
If we use atomic package to update this value, better to make this private and provide accessor using atomic.Load*?
Or use package like https://pkg.go.dev/go.uber.org/atomic ?

Reply:
This field is private until it is done updating. We don't update it after we publish it.

Reply:
I think that won't be matter in practice, but https://github.com/golang/go/issues/5045#issuecomment-252730563 says
> you shouldn't mix atomic and non-atomic accesses for a given memory word.

Reply:
yes, it says that and we indeed must follow those rules for concurrent reads and writes. But this is not the case here. 

To be be clear: uploader is a private struct. Other packages do not have access to it. Once we return the stat, we no longer write.

Reply:
I think copying value by return is not guaranteed to be valid operation too if atomic is used there?

Reply:
I don't think so. Do you have links to some documentation to backup this up?

Reply:
Seems like this is the only outstanding comment and I am pretty sure this code is correct. I am happy to fix it in another CL if it is not.